power_grid_model_io_native_c/power_grid_model_io_native/include/power_grid_model_io_native/vnf_converter/vnf_pgm_converter.hpp

```diff
@@ -0,0 +1,128 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+#ifndef POWER_GRID_MODEL_IO_NATIVE_C_VNF_PGM_CONVERTER_HPP
+#define POWER_GRID_MODEL_IO_NATIVE_C_VNF_PGM_CONVERTER_HPP
+
+#include <power_grid_model/auxiliary/dataset.hpp>
+#include <power_grid_model/common/exception.hpp>
+
+#include <iostream>
+
+class PgmVnfConverter {
+  public:
+    PgmVnfConverter(char* buffer = nullptr, power_grid_model::WritableDataset* data = nullptr);
+
+    // Public member functions
+    void parse_vnf_file();
+    power_grid_model::ConstDataset const* convert_input(power_grid_model::ConstDataset const* dataset);
+
+  private:
+    // Private attributes
+    char* f_file_buffer;
+    power_grid_model::WritableDataset*
+        deserialized_data; // this type because it is generated by a deserializer type structure
+
+    // Private setters/getters
+    void set_file_buffer(char* file_buffer);
+
+    void set_deserialized_data(power_grid_model::WritableDataset* deserialized_data);
+
+    char* get_file_buffer();
+
+    power_grid_model::WritableDataset* get_deserialized_data();
+
+    // Private member functions
+    void convert_node_input();
+    void convert_line_input();
+    void convert_sources_input();
+    void convert_sym_loads_input();
+    void convert_shunts_input();
+    void convert_transformer_input();
+    void convert_sym_gens_input();
+    void convert_links_input();
+};
+
+inline PgmVnfConverter::PgmVnfConverter(char* buffer, power_grid_model::WritableDataset* data)
```

Comment:
is there a particular reason why you went with out-of-line implementations instead of in-class implementations (as we do in the PGM):

```cpp
class PgmVnfConverter {
  public:
    PgmVnfConverter(char* buffer = nullptr, power_grid_model::WritableDataset* data = nullptr) 
        : f_file_buffer(buffer), deserialized_data(data) {
            using namespace std::string_literals;
            using power_grid_model::ExperimentalFeature;
            throw ExperimentalFeature{"PGM_VNF_converter", ExperimentalFeature::TypeValuePair{.name = "PGM_VNF_conversion",
                                                                                              .value = std::to_string(1)}};
    }

    // the rest of the functions
    // ...
};
```